File: lib-segment/src/main/java/websiteschema/mpsegment/filter/UpperCaseAndHalfShapeFilter.java
package websiteschema.mpsegment.filter;

import websiteschema.mpsegment.util.StringUtil;

public class UpperCaseAndHalfShapeFilter extends AbstractSegmentFilter {

    public UpperCaseAndHalfShapeFilter(boolean halfShapeAll, boolean upperCaseAll) {
        this.halfShapeAll = halfShapeAll;
        this.upperCaseAll = upperCaseAll;
        this.upperCaseOrHalfShapeAll = halfShapeAll || upperCaseAll;
    }

    @Override
    public void doFilter() {
        int length = segmentResult.length();
        for (int index = 0; index < length; index++) {
            String word = segmentResult.getWord(index);
            segmentResult.setWord(index, filterWord(word));
        }
    }

    private String filterWord(String word) {
        if (upperCaseOrHalfShapeAll) {
            if (halfShapeAll && upperCaseAll) {
                return StringUtil.doUpperCaseAndHalfShape(word);
            }
            if (halfShapeAll) {
                return StringUtil.halfShape(word);
            }
            return StringUtil.toUpperCase(word);
        }
        return word;
    }

    private boolean upperCaseOrHalfShapeAll;
    private boolean upperCaseAll;
    private boolean halfShapeAll;
}

File: lib-segment/src/main/scala/me/yingrui/segment/filter/UpperCaseAndHalfShapeFilter.scala
package me.yingrui.segment.filter

import me.yingrui.segment.util.StringUtil

class UpperCaseAndHalfShapeFilter(halfShapeAll: Boolean, upperCaseAll: Boolean) extends AbstractSegmentFilter {

  private val upperCaseOrHalfShapeAll = halfShapeAll || upperCaseAll

  override def doFilter() {
    val length = segmentResult.length()
    for (index <- 0 until length) {
      val word = segmentResult.getWord(index)
      segmentResult.setWord(index, filterWord(word))
    }
  }

  private def filterWord(word: String): String = {
    if (upperCaseOrHalfShapeAll) {
      if (halfShapeAll && upperCaseAll) {
        return StringUtil.doUpperCaseAndHalfShape(word)
      }
      if (halfShapeAll) {
        return StringUtil.halfShape(word)
      }
      return StringUtil.toUpperCase(word)
    }
    return word
  }

}
